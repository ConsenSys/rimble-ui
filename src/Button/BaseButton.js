import React from 'react';
import styled from 'styled-components';
import {
  color,
  space,
  minWidth,
  fontSize,
  fontFamily,
  fontWeight,
  buttonStyle,
  variant,
  style,
  themeGet,
} from 'styled-system';

import defaultTheme from '../theme';
import Box from '../Box';
import Icon from '../Icon';

const buttonSize = variant({
  prop: 'size',
  key: 'buttonSizes',
});

const mainColor = style({
  prop: 'mainColor',
  cssProperty: '--main-color',
  key: 'colors',
});

const contrastColor = style({
  prop: 'contrastColor',
  cssProperty: '--contrast-color',
  key: 'colors',
});

// const fullWidth = props => (props.fullWidth ? { width: '100%' } : null);

const StyledButton = styled(Box)`
  ${mainColor}
  ${contrastColor}

  & {
    -webkit-font-smoothing: antialiased;
    appearance: none;
    cursor: pointer;
    overflow: hidden;
    white-space: nowrap;
    text-decoration: none;
    text-align: center;

    position: relative;
    display: inline-flex;
    align-items: center;
    justify-content: center;

    padding: ${props => (props.icononly ? '0' : props.p)};
  }

  ${'' /* with icon styles */}
  > svg,
  > span.button-text {
    position: relative;
    z-index: 1;
  }

  > svg {
    display: block;
  }
  > svg:first-child {
    margin-left: -0.5rem;
    margin-right: .5rem;
  }
  > svg:last-child {
    margin-left: .5rem;
    margin-right: -0.5rem;
  }
  > svg:first-child:last-child {
    margin: 0;
  }

  ${'' /* styled system props */}
  ${color}
  ${space}
  ${minWidth}

  ${fontSize}
  ${fontFamily}
  ${fontWeight}

  ${buttonStyle}
  ${buttonSize}

  ${'' /* ${boxShadow} */}
  ${'' /* ${opacity} */}

  ${'' /* ${fullWidth} */}
`;

const Button = ({ className, children, ...props }) => {
  if (props.icon) {
    return (
      <StyledButton type="button" className={className} {...props}>
        {props.icon && !props.iconpos && <Icon name={props.icon} />}
        {props.icon && props.iconpos === 'left' && <Icon name={props.icon} />}
        {children && <span className="button-text">{children}</span>}
        {props.icon && props.iconpos === 'right' && <Icon name={props.icon} />}
      </StyledButton>
    );
  }
  return (
    <StyledButton as="button" type="button" className={className} {...props}>
      <span className="button-text">{children}</span>
    </StyledButton>
  );
};

Button.defaultProps = {
  // base props
  theme: defaultTheme,
  as: 'button',
  // variant: 'primary',
  size: 'medium',
  m: 0,
  px: 4,
  py: 0,
  fontSize: 'inherit',
  fontFamily: 'sansSerif',
  fontWeight: 3,
  // color props

  color: 'white',
  bg: 'silver',

  mainColor: 'primary',
  contrastColor: 'white',

  // height: '48px',

  border: 'none',
  // borderColor: 'none',
  // borderRadius: 1,
};

Button.displayName = 'Button';

export default Button;
