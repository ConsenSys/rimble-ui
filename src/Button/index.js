import React from 'react';
import styled from 'styled-components';
import {
  fontFamily,
  fontWeight,
  fontSize,
  boxShadow,
  opacity,
  themeGet,
} from 'styled-system';

import { tint, shade } from 'polished';

import Box from '../Box';
import Icon from '../Icon';

import defaultTheme from '../theme';

const fullWidth = props => (props.fullWidth ? { width: '100%' } : null);

const size = props => {
  switch (props.size) {
    case 'small':
      return `
        font-size: 12px;
        height: 2rem;
        min-width: 2rem;
        padding: 0 1rem;
      `;
    case 'medium':
      return `
        font-size: 16px;
        height: 3rem;
        min-width: 3rem;
      `;
    case 'large':
      return `
        font-size: 24px;
        height: 4rem;
        min-width: 4rem;
      `;
    default:
      return `
        font-size: 1rem;
        min-width: 3rem;
      `;
  }
};

const ProtoButton = React.forwardRef((props, ref) => (
  <Box as="button" type="button" ref={ref} {...props}>
    {props.icon && !props.iconpos && <Icon name={props.icon} />}
    {props.icon && props.iconpos === 'left' && <Icon name={props.icon} />}
    {props.children && <span>{props.children}</span>}
    {props.icon && props.iconpos === 'right' && <Icon name={props.icon} />}
  </Box>
));

const Button = styled(ProtoButton).attrs({
  hovercolor: props => themeGet('colors.primary', 'black')(props),
})`
  & {
    cursor: pointer;
    text-decoration: none;
    text-align: center;
    line-height: 1;
    overflow: hidden;
    white-space: nowrap;
    -webkit-font-smoothing: antialiased;

    display: inline-flex;
    align-items: center;
    justify-content: center;
    transition: all .15s ease;

    padding: ${props => (props.icononly ? '0' : props.p)};
    background-color: ${props => props.hovercolor};
  }
  &:hover {
    background-color: ${props => tint(0.1, props.hovercolor)};
    box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.2);
  }
  &:active {
    background-color: ${props => shade(0.1, props.hovercolor)};
    box-shadow: none;
  }
  &:disabled {
    opacity: ${props => themeGet('opacity.disabled', '0.5')};
    cursor: not-allowed;
  }

  > svg {
    display: block;
  }
  > svg:first-child {
    margin-left: -1rem;
    margin-right: .5rem;
  }
  > svg:last-child {
    margin-left: .5rem;
    margin-right: -1rem;
  }
  > svg:first-child:last-child {
    margin: 0;
  }

  ${fontSize}
  ${fontWeight}
  ${fontFamily}
  ${boxShadow}
  ${opacity}

  ${size}
  ${fullWidth}
`;

Button.defaultProps = {
  theme: defaultTheme,
  position: 'relative',
  color: 'white',
  bg: 'primary',
  height: '48px',
  m: 0,
  px: 4,
  py: 0,
  border: 0,
  borderColor: 'none',
  borderRadius: 1,
  boxShadow: 1,
  fontSize: 'inherit',
  fontFamily: 'sansSerif',
  fontWeight: 3,
};

Button.displayName = 'Button';

export default Button;
