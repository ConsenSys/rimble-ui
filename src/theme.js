import { tint, shade } from 'polished';

const success = '#28C081';
const danger = '#DC2C10';
const warning = '#FD9D28';
const info = '#36ADF1';

// theme.js
export default {
  fontSizes: [12, 14, 16, 20, 24, 32, 48, 64],
  fontWeights: [0, 300, 400, 600, 700],
  letterSpacings: [0, 1, 2, 4, 8],
  lineHeights: {
    solid: 1,
    title: 1.25,
    copy: 1.5,
  },
  fonts: {
    serif: 'athelas, georgia, times, serif',
    sansSerif: '"Source Sans Pro", -apple-system, sans-serif',
  },
  space: [0, 4, 8, 16, 32, 64, 128, 256],
  radii: ['0', '4px', '8px', '16px'],
  width: [0, 16, 32, 64, 128, 256],
  minWidths: [0, 16, 32, 64, 128, 256],
  maxWidths: [0, 16, 32, 64, 128, 256, 512, 768, 1024, 1536],
  heights: [0, 16, 32, 64, 128, 256],
  minHeights: [0, 16, 32, 64, 128, 256],
  maxHeights: [0, 16, 32, 64, 128, 256],
  borders: [0, '1px solid transparent'],
  borderWidths: ['0', '1px', '2px', '4px'],
  shadows: [
    '0',
    '0px 2px 4px rgba(0, 0, 0, 0.1)',
    '0 7px 14px rgba(50,50,93,.1)',
  ],
  opacity: {
    disabled: 0.4,
  },
  colors: {
    primary: '#4E3FCE',
    blue: '#007ce0',
    navy: '#004175',
    copyColor: '#3F3D4B',
    // black: '#000',
    black: '#000e1a',
    'near-black': '#111',
    'dark-gray': '#333',
    'mid-gray': '#555',
    // gray: ' #777',
    grey: '#CCC',
    silver: '#999',
    'light-silver': '#aaa',
    'moon-gray': '#ccc',
    'light-gray': '#eee',
    'near-white': '#f4f4f4',
    white: '#fff',
    transparent: 'transparent',
    blacks: [
      'rgba(0,0,0,.0125)',
      'rgba(0,0,0,.025)',
      'rgba(0,0,0,.05)',
      'rgba(0,0,0,.1)',
      'rgba(0,0,0,.2)',
      'rgba(0,0,0,.3)',
      'rgba(0,0,0,.4)',
      'rgba(0,0,0,.5)',
      'rgba(0,0,0,.6)',
      'rgba(0,0,0,.7)',
      'rgba(0,0,0,.8)',
      'rgba(0,0,0,.9)',
    ],
    whites: [
      'rgba(255,255,255,.0125)',
      'rgba(255,255,255,.025)',
      'rgba(255,255,255,.05)',
      'rgba(255,255,255,.1)',
      'rgba(255,255,255,.2)',
      'rgba(255,255,255,.3)',
      'rgba(255,255,255,.4)',
      'rgba(255,255,255,.5)',
      'rgba(255,255,255,.6)',
      'rgba(255,255,255,.7)',
      'rgba(255,255,255,.8)',
      'rgba(255,255,255,.9)',
    ],
    danger: danger,
    warning: warning,
    success: success,
    info: info,
  },
  zIndices: [0, 9, 99, 999, 9999],
  messageStyle: {
    base: {
      color: shade(0.4, '#AAA'),
      backgroundColor: tint(0.9, '#AAA'),
      borderColor: '#AAA',
    },
    success: {
      color: shade(0.4, success),
      backgroundColor: tint(0.9, success),
      borderColor: success,
    },
    warning: {
      color: shade(0.4, warning),
      backgroundColor: tint(0.9, warning),
      borderColor: warning,
    },
    danger: {
      color: shade(0.4, danger),
      backgroundColor: tint(0.9, danger),
      borderColor: danger,
    },
    info: {
      color: shade(0.4, info),
      backgroundColor: tint(0.9, info),
      borderColor: info,
    },
  },
};
