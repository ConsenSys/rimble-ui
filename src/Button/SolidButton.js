import React from 'react';
import styled from 'styled-components';

import { boxShadow } from 'styled-system';

import Button from './BaseButton';

// outline
const StyledSolidButton = styled(Button)`
  & {
    color: var(--contrast-color);
    ${'' /* background: var(--main-color); */}
    background: #FFF;

    position: relative;
    z-index: 0;
    transition: all 0.15s ease;
  }

  &:hover {
    text-decoration: inherit;
    box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.2);
    ${'' /* opacity: 0.5; */}
  }

  &:active {
    box-shadow: none;
    background: #000;
  }

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }

  &::before {
    content: '';
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 1;
    height: 100%;
    width: 100%;
    ${'' /* opacity: 0; */}
    ${'' /* background: none; */}
    background: var(--main-color);
    ${'' /* background: #fff; */}
    pointer-events: none;
  }

  &:hover::before {
    ${'' /* background: #fff; */}
    opacity: 0.9;
  }

  ${'' /* &:active::before {
    background: #000;
    opacity: 0.1;
  } */}

  ${boxShadow}
`;

const SolidButton = ({ className, children, ...props }) => {
  return (
    <StyledSolidButton className={className} {...props}>
      {children}
    </StyledSolidButton>
  );
};

SolidButton.defaultProps = {
  borderRadius: 1,
  boxShadow: 1,
};

export default SolidButton;
