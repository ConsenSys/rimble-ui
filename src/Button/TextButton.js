import React from 'react';
import styled from 'styled-components';

import Button from './BaseButton';

// text
const StyledTextButton = styled(Button)`
  & {
    color: var(--main-color);
    border: none;
    background: none;
    box-shadow: none;
  }
  &:hover {
    text-decoration: underline;
    background: inherit;
    box-shadow: inherit;
  }
  &:active {
    text-decoration: none;
  }
  &::before {
    display: none;
  }
`;

const TextButton = ({ className, children, ...props }) => {
  return (
    <StyledTextButton className={className} {...props}>
      {children}
    </StyledTextButton>
  );
};

TextButton.defaultProps = {
  px: 2,
};

export default TextButton;
