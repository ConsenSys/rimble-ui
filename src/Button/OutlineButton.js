import React from 'react';
import styled from 'styled-components';

import SolidButton from './SolidButton';

// outline
const StyledOutlineButton = styled(SolidButton)`
  & {
    color: var(--main-color);
    border: 1px solid;
    border-color: gray;
    background: transparent;
  }

  &&:hover {
    border-color: var(--main-color);
  }

  ${'' /* &&:hover::before {
    opacity: 0;
    background-color: var(--main-color);
  } */}

  &&:active::before {
    opacity: 0.1;
    background: var(--main-color);
  }
`;

const OutlineButton = ({ className, children, ...props }) => {
  return (
    <StyledOutlineButton className={className} {...props}>
      {children}
    </StyledOutlineButton>
  );
};

export default OutlineButton;
