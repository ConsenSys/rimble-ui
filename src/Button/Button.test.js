import Button from './'
import ThemeProvider from '../ThemeProvider'
import theme from '../theme'

describe('Button component sanity', () => {
  it('has name', () => {
    expect(Button.displayName).toBe('Button')
  })

  it('matches default snapshot', () => {
    const component = render(
      <Button />
    )
    expect(component).toMatchSnapshot()
  })

  it('matches themed snapshot', () => {
    const component = render(
      <ThemeProvider theme={theme}>
        <Button />
      </ThemeProvider>
    )
    expect(component).toMatchSnapshot()
  })
})
