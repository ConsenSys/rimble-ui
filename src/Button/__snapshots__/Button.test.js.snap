// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component sanity matches default snapshot 1`] = `
.c1 {
  margin: 0;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 32px;
  padding-right: 32px;
  color: #fff;
  background-color: #4E3FCE;
  font-size: inherit;
  height: 48px;
  position: relative;
  border: 0;
  border-color: none;
  border-radius: 4px;
  border-color: none;
  border-radius: 4px;
  text-align: left;
}

.c1 {
  box-sizing: border-box;
}

.c0 {
  font-size: inherit;
  font-weight: 600;
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
  font-size: 1rem;
  min-width: 3rem;
}

.c0 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  line-height: 1;
  overflow: hidden;
  white-space: nowrap;
  -webkit-font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all .15s ease;
  transition: all .15s ease;
}

.c0.c0:hover {
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  background-color: #5f52d2;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.2);
}

.c0:active {
  background-color: #4638b9;
  box-shadow: none;
}

.c0:disabled {
  opacity: 0.4;
  cursor: not-allowed;
}

.c0 > svg {
  display: block;
}

.c0 > svg:first-child {
  margin-left: -1rem;
  margin-right: .5rem;
}

.c0 > svg:last-child {
  margin-left: .5rem;
  margin-right: -1rem;
}

.c0 > svg:first-child:last-child {
  margin: 0;
}

<button
  class="c0 c1"
  color="white"
  font-family="sansSerif"
  font-size="inherit"
  font-weight="3"
  height="48px"
  type="button"
>
  
</button>
`;

exports[`Button component sanity matches themed snapshot 1`] = `
.c2 {
  margin: 0;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 32px;
  padding-right: 32px;
  color: #fff;
  background-color: #4E3FCE;
  font-size: inherit;
  height: 48px;
  position: relative;
  border: 0;
  border-color: none;
  border-radius: 4px;
  border-color: none;
  border-radius: 4px;
  text-align: left;
}

.c2 {
  box-sizing: border-box;
}

.c1 {
  font-size: inherit;
  font-weight: 600;
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
  font-size: 1rem;
  min-width: 3rem;
}

.c1 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  line-height: 1;
  overflow: hidden;
  white-space: nowrap;
  -webkit-font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all .15s ease;
  transition: all .15s ease;
}

.c1.c1:hover {
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  background-color: #5f52d2;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.2);
}

.c1:active {
  background-color: #4638b9;
  box-shadow: none;
}

.c1:disabled {
  opacity: 0.4;
  cursor: not-allowed;
}

.c1 > svg {
  display: block;
}

.c1 > svg:first-child {
  margin-left: -1rem;
  margin-right: .5rem;
}

.c1 > svg:last-child {
  margin-left: .5rem;
  margin-right: -1rem;
}

.c1 > svg:first-child:last-child {
  margin: 0;
}

.c0 {
  font-size: 1em;
  font-family: "Source Sans Pro",-apple-system,sans-serif;
  color: #3F3D4B;
}

<div
  class="c0"
>
  <button
    class="c1 c2"
    color="white"
    font-family="sansSerif"
    font-size="inherit"
    font-weight="3"
    height="48px"
    type="button"
  >
    
  </button>
</div>
`;
